### PR TITLE
[LLM] Unify Langchain Native and Transformers LLM API

### PR DESCRIPTION
**Pull request by sgwhat**
## Description
This PR primarily aims to unify the API Usage difference between BigDL-LLM langchain transformers and native optimization.

### 1. Why the change?

Align with https://github.com/intel-analytics/BigDL/pull/8713, see https://github.com/analytics-zoo/nano/issues/508 for more details.

### 2. User API changes
- `langchain.llms`
```
from langchain.llms import LlamaLLM

# Native Int4
llm = LlamaLLM(model_path="/path/to/llama_int4.bin", native=True)

# Transformers Int4
llm = LlamaLLM(model_path="/path/to/llama_hf", native=False)
```

- `langchain.embeddings`
```
from bigdl.llm.langchain.embeddings import LlamaEmbedding

# Native Int4
llama = LlamaEmbedding(model_path="/path/to/llama_int4.bin", native=True)

# Transformers Int4
llama = LlamaEmbedding(model_path="/path/to/llama_hf", native=False)
```

### 3. Summary of the change 
- Deprecate `BigdlNativeEmbeddings` and `BigdlNativeLLM`.
- For `langchain.llms`, we added `LlamaLLM`, etc.
- For `langchain.embeddings`, we added `LlamaLMEmbeddings`, etc.

### 4. How to test?
- [x] Unit test
- [x] Application test

Chatglm ut and readthedocs updating will be added in another separate PR.

### CODE REVIEW

**Comment by sgwhat** [comment]
Chatglm ut and readthedocs updating will be added in another separate PR.

File: python/llm/README.md [hunk 1]
```diff
@@ -148,18 +148,24 @@ You may run the models using the LangChain API in `bigdl-llm`.
...
+  from bigdl.llm.langchain.llms import LlamaLLM
+  from bigdl.llm.langchain.embeddings import LlamaEmbeddings
   from langchain.chains.question_answering import load_qa_chain
 
-  embeddings = BigdlNativeEmbeddings(model_path='/path/to/converted/model.bin',
-                            model_family="llama",...)
-  bigdl_llm = BigdlNativeLLM(model_path='/path/to/converted/model.bin',
+  #switch to ChatGLMEmbeddings/GptneoxEmbeddings/BloomEmbeddings/StarcoderEmbeddings to load other models
+  embeddings = LlamaEmbeddings(model_path='/path/to/converted/model.bin',
+                               model_family="llama",...)
```
Comment (hkvision) [inline python/llm/README.md:166]: why LlamaEmbeddings still need to have the argument `model_family="llama"`?

**Comment by jason-dai** [comment]
To make it clear, add a `native` argument to the API (similar to `LlamaForCausalLM`); you may require it to be true for now.

**Comment by sgwhat** [comment]
> To make it clear, add a `native` argument to the API (similar to `LlamaForCausalLM`); you may require it to be true for now.

For sure we had added `native` to this API which defaulted to be true, and it will be modified to be a required argument.

File: python/llm/README.md (hunk 1 above)
Comment (sgwhat) [reply python/llm/README.md]: Fixed.

File: python/llm/example/langchain/native_int4/docqa.py [hunk 2]
```diff
@@ -31,16 +31,14 @@
...
-
+from bigdl.llm.langchain.llms import LlamaLLM
+from bigdl.llm.langchain.embeddings import LlamaEmbeddings
 
 
 def main(args):
     
     input_path = args.input_path 
     model_path = args.model_path 
-    model_family = args.model_family
```
Comment (hkvision) [inline python/llm/example/langchain/native_int4/docqa.py:43]: in the example, I think we still need this argument to support different models.

**Comment by hkvision** [comment]
Others LGTM

File: python/llm/example/langchain/native_int4/docqa.py (hunk 2 above)
Comment (hkvision) [reply python/llm/example/langchain/native_int4/docqa.py:43]: also for other examples.

File: python/llm/README.md [hunk 3]
```diff
@@ -148,19 +148,23 @@ You may run the models using the LangChain API in `bigdl-llm`.
  
 - **Using native INT4 format**
 
-  You may also convert Hugging Face *Transformers* models into *native INT4* format (currently only *llama*/*bloom*/*gptneox*/*starcoder* model family is supported), and then run the converted models using the LangChain API as follows.
+  You may also convert Hugging Face *Transformers* models into *native INT4* format, and then run the converted models using the LangChain API as follows.
   
-  >**Note**: Currently only llama/bloom/gptneox/starcoder model family is supported; for other models, you may use the Transformers INT4 format as described above).
+  >**Notes**: 
+  
+  * Currently only llama/bloom/gptneox/starcoder/chatglm model families are supported; for other models, you may use the Transformers INT4 format as described above).
```
Comment (hkvision) [inline python/llm/README.md:155]: Transformers INT4 -> Hugging Face `transformers` INT4

File: python/llm/example/langchain/native_int4/docqa.py [hunk 4]
```diff
@@ -54,15 +52,39 @@ def main(args):
...
+
+    if model_family in model_family_to_embeddings and model_family in model_family_to_llm:
+        llm_embeddings = model_family_to_embeddings[model_family]
+        langchain_llm = model_family_to_llm[model_family]
+    else:
+        raise ValueError(f"Unknown model family: {model_family}")
+
     # create embeddings and store into vectordb
-    embeddings = BigdlNativeEmbeddings(model_path=model_path, model_family=model_family, n_threads=n_threads, n_ctx=n_ctx)
+    # switch to ChatGLMEmbeddings/GptneoxEmbeddings/BloomEmbeddings/StarcoderEmbeddings to load other models
```
Comment (hkvision) [inline python/llm/example/langchain/native_int4/docqa.py:78]: remove this

File: python/llm/example/langchain/native_int4/docqa.py [hunk 5]
```diff
@@ -54,15 +52,39 @@ def main(args):
...
     docsearch = Chroma.from_texts(texts, embeddings, metadatas=[{"source": str(i)} for i in range(len(texts))]).as_retriever()
 
-    #get relavant texts
+    # get relavant texts
     docs = docsearch.get_relevant_documents(query)
-        
-    bigdl_llm = BigdlNativeLLM(
-        model_path=model_path, model_family=model_family, n_ctx=n_ctx, n_threads=n_threads, callback_manager=callback_manager
+
+    # switch to ChatGLMLLM/GptneoxLLM/BloomLLM/StarcoderLLM to load other models
```
Comment (hkvision) [inline python/llm/example/langchain/native_int4/docqa.py:85]: remove this, also for other examples

File: python/llm/src/bigdl/llm/langchain/embeddings/bigdlllm.py [hunk 6]
```diff
@@ -190,3 +200,191 @@ def embed_query(self, text: str) -> List[float]:
         """
         embedding = self.client.embed(text)
         return list(map(float, embedding))
+
+
+class _BaseEmbeddings(BaseModel, Embeddings):
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/langchain/embeddings/bigdlllm.py:205]: Can BigdlNativeEmbeddings merge the code with this class? Currently too much duplicate code

Comment (sgwhat) [reply python/llm/src/bigdl/llm/langchain/embeddings/bigdlllm.py:205]: I don't think so :( .

**Review by hkvision — APPROVED** [review]

**Pull request merged**